File: packages/go/cairn-p2p/crypto/aead.go
package crypto

import (
	"crypto/aes"
	"crypto/cipher"
	"fmt"

	"golang.org/x/crypto/chacha20poly1305"
)

const (
	// NonceSize is the nonce size for both AEAD ciphers: 12 bytes.
	NonceSize = 12
	// KeySize is the key size for both AEAD ciphers: 32 bytes.
	KeySize = 32
	// TagSize is the authentication tag size for both ciphers: 16 bytes.
	TagSize = 16
)

// CipherSuite identifies an AEAD cipher.
type CipherSuite int

const (
	CipherAes256Gcm        CipherSuite = iota
	CipherChaCha20Poly1305
)

// AeadEncrypt encrypts plaintext with associated data using the specified cipher.
// Returns ciphertext with appended authentication tag.
func AeadEncrypt(cs CipherSuite, key [32]byte, nonce [12]byte, plaintext, aad []byte) ([]byte, error) {
	aead, err := newAEAD(cs, key)
	if err != nil {
		return nil, err
	}
	ciphertext := aead.Seal(nil, nonce[:], plaintext, aad)
	return ciphertext, nil
}

// AeadDecrypt decrypts ciphertext with associated data using the specified cipher.
// Returns plaintext on success, or error if authentication fails.
func AeadDecrypt(cs CipherSuite, key [32]byte, nonce [12]byte, ciphertext, aad []byte) ([]byte, error) {
	aead, err := newAEAD(cs, key)
	if err != nil {
		return nil, err
	}
	plaintext, err := aead.Open(nil, nonce[:], ciphertext, aad)
	if err != nil {
		return nil, fmt.Errorf("AEAD decryption failed: %w", err)
	}
	return plaintext, nil
}

func newAEAD(cs CipherSuite, key [32]byte) (cipher.AEAD, error) {
	switch cs {
	case CipherAes256Gcm:
		block, err := aes.NewCipher(key[:])
		if err != nil {
			return nil, fmt.Errorf("AES cipher creation failed: %w", err)
		}
		aead, err := cipher.NewGCM(block)
		if err != nil {
			return nil, fmt.Errorf("AES-GCM creation failed: %w", err)
		}
		return aead, nil
	case CipherChaCha20Poly1305:
		aead, err := chacha20poly1305.New(key[:])
		if err != nil {
			return nil, fmt.Errorf("ChaCha20-Poly1305 creation failed: %w", err)
		}
		return aead, nil
	default:
		return nil, fmt.Errorf("unknown cipher suite: %d", cs)
	}
}

File: packages/ts/cairn-p2p/src/crypto/aead.ts
import { gcm } from '@noble/ciphers/aes';
import { ChaCha20Poly1305 } from '@stablelib/chacha20poly1305';
import { CairnError } from '../errors.js';
import type { CipherSuite } from '../config.js';

/** Nonce size for both ciphers: 12 bytes. */
export const NONCE_SIZE = 12;
/** Key size for both ciphers: 32 bytes. */
export const KEY_SIZE = 32;
/** Authentication tag size for both ciphers: 16 bytes. */
export const TAG_SIZE = 16;

/**
 * Encrypt plaintext with associated data using the specified cipher.
 *
 * Returns ciphertext with appended authentication tag.
 */
export function aeadEncrypt(
  cipher: CipherSuite,
  key: Uint8Array,
  nonce: Uint8Array,
  plaintext: Uint8Array,
  aad: Uint8Array,
): Uint8Array {
  if (key.length !== KEY_SIZE) {
    throw new CairnError('CRYPTO', `AEAD key must be ${KEY_SIZE} bytes, got ${key.length}`);
  }
  if (nonce.length !== NONCE_SIZE) {
    throw new CairnError('CRYPTO', `AEAD nonce must be ${NONCE_SIZE} bytes, got ${nonce.length}`);
  }

  try {
    if (cipher === 'aes-256-gcm') {
      const aes = gcm(key, nonce, aad);
      return aes.encrypt(plaintext);
    } else {
      const chacha = new ChaCha20Poly1305(key);
      return chacha.seal(nonce, plaintext, aad);
    }
  } catch (e) {
    throw new CairnError('CRYPTO', `AEAD encrypt error: ${e}`);
  }
}

/**
 * Decrypt ciphertext with associated data using the specified cipher.
 *
 * Returns plaintext on success. Throws CairnError if authentication fails.
 */
export function aeadDecrypt(
  cipher: CipherSuite,
  key: Uint8Array,
  nonce: Uint8Array,
  ciphertext: Uint8Array,
  aad: Uint8Array,
): Uint8Array {
  if (key.length !== KEY_SIZE) {
    throw new CairnError('CRYPTO', `AEAD key must be ${KEY_SIZE} bytes, got ${key.length}`);
  }
  if (nonce.length !== NONCE_SIZE) {
    throw new CairnError('CRYPTO', `AEAD nonce must be ${NONCE_SIZE} bytes, got ${nonce.length}`);
  }

  try {
    if (cipher === 'aes-256-gcm') {
      const aes = gcm(key, nonce, aad);
      return aes.decrypt(ciphertext);
    } else {
      const chacha = new ChaCha20Poly1305(key);
      const result = chacha.open(nonce, ciphertext, aad);
      if (result === null) {
        throw new CairnError('CRYPTO', 'ChaCha20-Poly1305 authentication failed');
      }
      return result;
    }
  } catch (e) {
    if (e instanceof CairnError) throw e;
    throw new CairnError('CRYPTO', `AEAD decrypt error: ${e}`);
  }
}
